### ISSUE TITLE
Cache invalidates types but data isn't refetched

### ISSUE THREAD

**Issue by michaeljwilliams** [FormidableLabs/urql#265]
I seem to be having a similar problem as in #195. After I submit a mutation, the types returned are invalidated in the cache, but when they are next queried no data is returned. They are not refetched and the query hangs with `fetching: true` and `data: undefined`. I can see that the network requests are never made.

For example, after this mutation
```gql
mutation editJob($input: InputEditJob!) {
    editJob(input: $input) {
        id
        name
        users {
            id
        }
        unitGroup {
            id
        }
    }
}
```

The following three queries hang:
```gql
query getJob($id: ID!) {
    getJobs(ids: [$id]) {
        id
        users {
            id
        }
        unitGroup {
            id
        }
    }
}
```
```gql
query getUnitGroups {
    getUnitGroups(ids: []) {
        id
        name
        comment
    }
}
```
```gql
query getUsers {
    getUsers(ids: []) {
        id
        username
    }
}
```

A sample response to the mutation:
```json
{
    "data": {
        "editJob": {
            "id": "34",
            "name": "new job",
            "users": [
                {
                    "id": "13",
                    "__typename": "User"
                }
            ],
            "unitGroup": {
                "id": "40",
                "__typename": "UnitGroup"
            },
            "__typename": "Job"
        },
        "__typename": "Mutation"
    }
}


```

I tried setting `requestPolicy='network-only'` on these queries but that had no effect. I turned off the cache exchange and everything works fine.

I'm currently using 1.1.0-0.alpha and I had the same issues in 1.0.5.

**Comment by kitten** [comment]
Hiya, I tried to replicate some of this scenario in our "getting started" example (`examples/1-getting-started/`) by adding two lists that both run different queries.

Unfortunately I'm not seeing any problem with refetching those queries after a mutation.

Can you please create some kind of reproduction or give some more information on what might be going on? A CodeSandbox or example repository would be incredibly helpful :+1:

**Comment by michaeljwilliams** [comment]
Extracting the relevant code from the app and creating a test server would take me a while... What information are you looking for? How can I see what the cache is doing and what's in it? I think that might help.

**Comment by kitten** [comment]
@michaeljwilliams The problem isn't really extracting the code but when you extract this stuff you're basically already narrowing down the code that might be responsible for the issue, which is what really saves time 👍

Regardless, let's see if we can figure this out quicker.

We have a debug exchange that you can use to observe what operations are ongoing. It'd be interesting to know whether you're also seeing some cancelled network requests in your devtools? Or is a new request not sent at all?

So, this also sounds like you have a component with a query, then unmount it, then invalidate it with a mutation, and lastly remount it?

Have you also checked the error property?

**Comment by michaeljwilliams** [comment]
There were no cancelled network requests and the error property was undefined.

I just tried to replicate the problem to get some logs, but somehow my issue has gone away. Here are the significant changes I made:

1. Added JWT authorization to the GraphQL endpoint.
2. Updated urql client to reflect authorization change.
3. Added an AppProvider global context component at the root of the component tree.
4. Moved UrqlProvider from being the parent of `<App />` to being a child of `<App />`.

When I did (2) I made the following changes:

From:
```js
const client = createClient({
    url: API_URI,
    fetchOptions: {
        mode: "cors", // no-cors, cors, *same-origin
        credentials: 'same-origin',
    },
})
```

To:
```js
const client = createClient({
    url: API_URI,
    fetchOptions: {
        mode: "cors", // no-cors, cors, *same-origin
        credentials: 'same-origin',
        contentType: 'application/json',
        headers: {
            authorization: AppState.auth.token,
        },
    },
})
```

When I did (3) and (4) I made the following changes:

From:
```jsx
// index.js (app entrypoint)
//...
render((
    <UrqlProvider value={client}>
        <App />
    </UrqlProvider>
), document.getElementById('root'))
```
To:
```jsx
// index.js (app entrypoint)
//...
render((
    <AppProvider>
        {/* I moved UrqlProvider into App. */}
        <App />
    </AppProvider>
), document.getElementById('root'))
```
```jsx
// App.js
//...
const App = () => {
    const {AppState, setAppState} = useContext(AppContext)

    const client = createClient({
        url: API_URI,
        fetchOptions: {
            mode: "cors", // no-cors, cors, *same-origin
            credentials: 'same-origin',
            contentType: 'application/json',
            headers: {
                authorization: AppState.auth.token,
            },
        },
    })
    return (
        <UrqlProvider value={client}>
            ...
        </UrqlProvider>
    )
}
```
Any ideas about what was happening?

Let me also add that I'm really enjoying working with urql. I started with Apollo Client and it was much too bloated and complicated. I like how lightweight, yet extensible urql is. Good work!

**Comment by michaeljwilliams** [comment]
@kitten Actually I've got another issue: Now it's not caching at all! :grin:  It's refetching data even when the exact same query is sent twice in a row. Should I open a new issue?

**Comment by kitten** [comment]
@michaeljwilliams I'm not sure whether this is possibly related to #280, but maybe you can give it a try? We've just [merged a fix](https://github.com/FormidableLabs/urql/pull/281#issue-286405338) and [published a patch version, v1.1.1](https://github.com/FormidableLabs/urql/releases/tag/v1.1.1)

**Comment by michaeljwilliams** [comment]
I updated to 1.1.1 and still no caching. I still see network requests for each query. Below are debug exchange outputs for two consecutive requests with the same query and variables (no variables).

[Exchange debug]: Incoming operation:
```json
{
  "key": 407556858,
  "query": {
    "kind": "Document",
    "definitions": [
      {
        "kind": "OperationDefinition",
        "operation": "query",
        "name": {
          "kind": "Name",
          "value": "getJobs"
        },
        "variableDefinitions": [],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "getJobs"
              },
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "ids"
                  },
                  "value": {
                    "kind": "ListValue",
                    "values": []
                  }
                }
              ],
              "directives": [],
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id"
                    },
                    "arguments": [],
                    "directives": []
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "name"
                    },
                    "arguments": [],
                    "directives": []
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "automation"
                    },
                    "arguments": [],
                    "directives": [],
                    "selectionSet": {
                      "kind": "SelectionSet",
                      "selections": [
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "id"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "name"
                          },
                          "arguments": [],
                          "directives": []
                        }
                      ]
                    }
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "unitGroup"
                    },
                    "arguments": [],
                    "directives": [],
                    "selectionSet": {
                      "kind": "SelectionSet",
                      "selections": [
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "id"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "name"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "status"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "quantity"
                          },
                          "arguments": [],
                          "directives": []
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ],
    "loc": {
      "start": 0,
      "end": 437,
      "source": {
        "body": "\n            query getJobs {\n                getJobs(ids: []) {\n                    id\n                    name\n                    automation {\n                        id\n                        name\n                    }\n                    unitGroup {\n                        id\n                        name\n                        status\n                        quantity\n                    }\n                }\n            }\n        ",
        "name": "GraphQL request",
        "locationOffset": {
          "line": 1,
          "column": 1
        }
      }
    },
    "__key": 4033129043
  },
  "variables": {},
  "operationName": "teardown",
  "context": {
    "url": "http://localhost:4000",
    "fetchOptions": {
      "mode": "cors",
      "credentials": "same-origin",
      "contentType": "application/json",
      "headers": {
        "authorization": "redacted"
      }
    },
    "requestPolicy": "cache-first"
  }
}
```
[Exchange debug]: Incoming operation:  
```json
{
  "key": 407556858,
  "query": {
    "kind": "Document",
    "definitions": [
      {
        "kind": "OperationDefinition",
        "operation": "query",
        "name": {
          "kind": "Name",
          "value": "getJobs"
        },
        "variableDefinitions": [],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "getJobs"
              },
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "ids"
                  },
                  "value": {
                    "kind": "ListValue",
                    "values": []
                  }
                }
              ],
              "directives": [],
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id"
                    },
                    "arguments": [],
                    "directives": []
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "name"
                    },
                    "arguments": [],
                    "directives": []
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "automation"
                    },
                    "arguments": [],
                    "directives": [],
                    "selectionSet": {
                      "kind": "SelectionSet",
                      "selections": [
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "id"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "name"
                          },
                          "arguments": [],
                          "directives": []
                        }
                      ]
                    }
                  },
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "unitGroup"
                    },
                    "arguments": [],
                    "directives": [],
                    "selectionSet": {
                      "kind": "SelectionSet",
                      "selections": [
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "id"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "name"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "status"
                          },
                          "arguments": [],
                          "directives": []
                        },
                        {
                          "kind": "Field",
                          "name": {
                            "kind": "Name",
                            "value": "quantity"
                          },
                          "arguments": [],
                          "directives": []
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ],
    "loc": {
      "start": 0,
      "end": 437,
      "source": {
        "body": "\n            query getJobs {\n                getJobs(ids: []) {\n                    id\n                    name\n                    automation {\n                        id\n                        name\n                    }\n                    unitGroup {\n                        id\n                        name\n                        status\n                        quantity\n                    }\n                }\n            }\n        ",
        "name": "GraphQL request",
        "locationOffset": {
          "line": 1,
          "column": 1
        }
      }
    },
    "__key": 4033129043
  },
  "variables": {},
  "operationName": "teardown",
  "context": {
    "url": "http://localhost:4000",
    "fetchOptions": {
      "mode": "cors",
      "credentials": "same-origin",
      "contentType": "application/json",
      "headers": {
        "authorization": "redacted"
      }
    },
    "requestPolicy": "cache-first"
  }
}
```
